### PR TITLE
NestedFolders: Refactor BrowseView state into redux

### PR DESCRIPTION
**Pull request by joshhunt**
In https://github.com/grafana/grafana/pull/66843 I said

> Now that there's a lot more state logic in here, and im pulling in immer myself, I think i want to store this in Redux instead.

All the state management in BrowseView was getting a bit too much, especially with the item selection logic. This PR moves this over to redux, which I think has made it much easier to grasp what's going on.

Related to https://github.com/grafana/grafana/issues/65604

### CODE REVIEW

**Comment by tskarhed** [comment]
Extracting the hooks made it easier to read, for sure. Overall I think it seems pretty manageable.

File: public/app/features/browse-dashboards/state/reducers.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+import { DashboardViewItem, DashboardViewItemKind } from 'app/features/search/types';
+
+import { BrowseDashboardsState } from '../types';
+
+import { fetchChildren } from './actions';
+
+type FetchChildrenAction = ReturnType<typeof fetchChildren.fulfilled>;
+
+export function extraReducerFetchChildrenFulfilled(state: BrowseDashboardsState, action: FetchChildrenAction) {
```
[the remark this answers is not in the document]
Comment (tskarhed) [reply public/app/features/browse-dashboards/state/reducers.ts:11]: I guess you'll do that when people think the approach is good :sweat_smile:

[the remark this answers is not in the document]
Comment (joshhunt) [reply public/app/features/browse-dashboards/state/reducers.ts:11]: wishful thinking!

But no. fair. Previously this was all tested by testing the view itself. Now that these are isolated, it is a lot easier to test them individually

File: public/app/features/browse-dashboards/state/reducers.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    while (nextParentUID) {
+      const parent = findItem(state.rootItems, state.childrenByParentUID, nextParentUID);
+      if (!parent) {
+        break;
+      }
+
+      state.selectedItems[parent.kind][parent.uid] = false;
+      nextParentUID = parent.parentUID;
+    }
```
Comment (ashharrison90) [inline public/app/features/browse-dashboards/state/reducers.ts:72]: i was struggling to get my head around this logic. had to play around with it in the UI before it sunk in. agree with @tskarhed that some tests or maybe even some comments to explain **why** we want to unselect here could really help clarify the behaviour 👍

**Review by ashharrison90 — APPROVED** [review]
this is really clean 😍 

only problems i had understanding anything was the actual selection logic, but that's because the logic is just naturally a little tricky.

great job 👏

File: public/app/features/browse-dashboards/state/reducers.ts (hunk 2 above)
Comment (joshhunt) [reply public/app/features/browse-dashboards/state/reducers.ts:72]: Yeah, fair. This is also kind of missing the other piece (which we've deprioritised) which is that the parent checkbox would change to a mixed/indeterminate to reflect that it's children are partially checked.

Comment (joshhunt) [reply public/app/features/browse-dashboards/state/reducers.ts]: added both unit tests, and some comments explaining this :)

lemme know how that goes - i understand both the feature and code is a little bit tricky to grok

**Review by ashharrison90 — APPROVED** [review]
looks great! 👏

**Pull request merged**